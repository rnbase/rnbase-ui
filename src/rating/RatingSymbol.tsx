import React, { useContext, useEffect, useState } from 'react';
import { Animated, StyleProp, TextProps, TextStyle } from 'react-native';
import { RatingContext, RatingValue } from './RatingContext';

export interface Props extends TextProps {
  animate: boolean;
  value: number;
  children?: React.ReactNode;
  style?: StyleProp<TextStyle>;
}

// let renders = 0;

const scale = (ratingValue: RatingValue, value: number) => (ratingValue.isEqual(value) ? 1.25 : 1);

const RatingSymbol: React.FC<Props> = ({ animate, value, children, style }) => {
  // console.log(`RatingSymbol, render(${++renders}): `, value);

  const ratingValue = useContext(RatingContext);

  const [animatedValue] = useState(() => new Animated.Value(scale(ratingValue, value)));

  useEffect(() => {
    if (animate) {
      return ratingValue.addListener(() => {
        Animated.spring(animatedValue, {
          tension: 50,
          friction: 3,
          toValue: scale(ratingValue, value),
          useNativeDriver: true,
        }).start();
      });
    }
  }, [animate, animatedValue, ratingValue, value]);

  return (
    <Animated.Text
      allowFontScaling={false}
      style={[style, { transform: [{ scale: animatedValue }] }]}
    >
      {children}
    </Animated.Text>
  );
};

export default RatingSymbol;
