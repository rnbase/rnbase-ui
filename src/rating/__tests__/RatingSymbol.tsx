import React from 'react';
import { Animated as AnimatedObj } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

import { Props } from '../RatingSymbol';

jest.doMock('react-native/Libraries/Animated/src/Animated', () => {
  const Animated = jest.requireActual('react-native/Libraries/Animated/src/Animated');

  return {
    ...Animated,
    spring: jest.fn((value: any, { toValue }: any) => {
      value.setValue(toValue);

      return {
        start: jest.fn(),
      };
    }),
  };
});

/**
 * Under test
 */
const { default: RatingSymbol } = jest.requireActual('../RatingSymbol');

const Animated: any = AnimatedObj;

const createElement = (props: Partial<Props>) => <RatingSymbol {...props} />;

const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  const tree = createRenderer({ children: '\u2605' });

  act(() => {});

  expect(tree).toMatchSnapshot();
});

it('should render with custom style', () => {
  const tree = createRenderer({
    children: '\u2605',
    style: { color: 'blue' },
  });

  act(() => {});

  expect(tree).toMatchSnapshot();
});

describe('animation', () => {
  const start = jest.fn();

  Animated.spring.mockImplementation(() => ({ start }));

  beforeEach(() => {
    start.mockClear();
  });

  it('should animate to selected', () => {
    const tree = createRenderer({ children: '\u2605' });

    act(() => {});

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ children: '\u2605', selected: true }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 1 }), {
      toValue: 1.25,
      friction: 3,
      tension: 50,
      useNativeDriver: true,
    });
    expect(start).toBeCalled();
  });

  it('should animate from selected', () => {
    const tree = createRenderer({ children: '\u2605', selected: true });

    act(() => {});

    Animated.spring.mockClear();

    act(() => {
      tree.update(createElement({ children: '\u2605' }));
    });

    expect(Animated.spring).toBeCalledWith(expect.objectContaining({ _value: 1.25 }), {
      toValue: 1,
      friction: 3,
      tension: 50,
      useNativeDriver: true,
    });
    expect(start).toBeCalled();
  });
});
