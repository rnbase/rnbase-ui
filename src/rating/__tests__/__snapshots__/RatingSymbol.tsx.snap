// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<Text
  allowFontScaling={false}
  style={
    Object {
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  ★
</Text>
`;

exports[`should render with custom style 1`] = `
<Text
  allowFontScaling={false}
  style={
    Object {
      "color": "blue",
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  ★
</Text>
`;
